File: bigdime-data-handlers/src/main/java/io/bigdime/handler/hive/LatencyBasedNextRunChecker.java
package io.bigdime.handler.hive;

import java.util.Map;
import java.util.concurrent.TimeUnit;

import io.bigdime.alert.LoggerFactory;
import io.bigdime.core.commons.AdaptorLogger;

public class LatencyBasedNextRunChecker extends AbstractNextRunChecker {
	private static final AdaptorLogger logger = new AdaptorLogger(
			LoggerFactory.getLogger(LatencyBasedNextRunChecker.class));
	private long intervalInMillis = TimeUnit.DAYS.toMillis(1);

	@Override
	public long getDateTimeInMillisForNextRun(long lastRunDateTime, final HiveJdbcReaderHandlerConfig handlerConfig,
			Map<? extends String, ? extends Object> properties) {
		long now = getAdjustedCurrentTime(handlerConfig.getLatency());

		long nextRunDateTime = 0;

		if (lastRunDateTime == 0) {// this is the first time
			return getDateTimeInMillisForFirstRun(handlerConfig, now);
		} else {
			nextRunDateTime = getDateTimeInMillisForSubsequentRun(handlerConfig, now, lastRunDateTime);
		}
		return nextRunDateTime;
	}

	protected long getDateTimeInMillisForSubsequentRun(final HiveJdbcReaderHandlerConfig handlerConfig, long now,
			long lastRunDateTime) {
		long nextRunDateTime = 0;
		if (now - lastRunDateTime > handlerConfig.getMinGoBack()) {

			nextRunDateTime = lastRunDateTime + intervalInMillis;
			logger.info("getDateTimeInMillisForNextRun",
					"_message=\"time to set hiveConfDateTime.\" now={} lastRunDateTime={} hiveConfDateTime={} intervalInMillis={}",
					now, lastRunDateTime, nextRunDateTime, intervalInMillis);

		} else {
			logger.info("nothing to do", "now={} hiveConfDateTime={} intervalInMillis={} time_until_next_run={}", now,
					lastRunDateTime, intervalInMillis, handlerConfig.getMinGoBack() - (now - lastRunDateTime));
		}
		return nextRunDateTime;
	}

}

File: bigdime-data-handlers/src/main/scala/io/bigdime/handler/hive/LatencyBasedNextRunChecker.scala
package io.bigdime.handler.hive

import java.util
import java.util.concurrent.TimeUnit

import io.bigdime.alert.LoggerFactory
import io.bigdime.core.commons.AdaptorLogger

/**
  * Created by neejain on 2/6/17.
  */
object LatencyBasedNextRunChecker {
  private val logger = new AdaptorLogger(LoggerFactory.getLogger(classOf[LatencyBasedNextRunChecker]))
  private val intervalInMillis = TimeUnit.DAYS.toMillis(1)
}

case class LatencyBasedNextRunChecker() extends AbstractNextRunChecker {

  import LatencyBasedNextRunChecker._

  override def getDateTimeInMillisForNextRun(lastRunDateTime: Long, handlerConfig: HiveJdbcReaderHandlerConfig, properties: util.Map[_ <: String, _]): Long = {
    val now = getAdjustedCurrentTime(handlerConfig.getLatency)
    if (lastRunDateTime == 0) {
      // this is the first time
      getDateTimeInMillisForFirstRun(handlerConfig, now)
    } else getDateTimeInMillisForSubsequentRun(handlerConfig, now, lastRunDateTime)
  }

  protected def getDateTimeInMillisForSubsequentRun(handlerConfig: HiveJdbcReaderHandlerConfig, now: Long, lastRunDateTime: Long): Long = {
    var nextRunDateTime = 0l
    if (now - lastRunDateTime > handlerConfig.getMinGoBack) {
      nextRunDateTime = lastRunDateTime + intervalInMillis
      logger.info("getDateTimeInMillisForNextRun", "_message=\"time to set hiveConfDateTime.\" now={} lastRunDateTime={} hiveConfDateTime={} intervalInMillis={}", now: java.lang.Long, lastRunDateTime: java.lang.Long, nextRunDateTime: java.lang.Long, intervalInMillis: java.lang.Long)
    }
    else logger.info("nothing to do", "now={} hiveConfDateTime={} intervalInMillis={} time_until_next_run={}", now: java.lang.Long, lastRunDateTime: java.lang.Long, intervalInMillis: java.lang.Long, (handlerConfig.getMinGoBack - (now - lastRunDateTime)): java.lang.Long)
    nextRunDateTime
  }
}
